### PR TITLE
fix: always process pending endorsements

### PR DESCRIPTION
**Pull request by Longarithm**
As I shortly described in https://near.zulipchat.com/#narrow/stream/407237-core.2Fstateless-validation/topic/chunk.20completion.20not.20guaranteed.20on.20BPs/near/441424842, there was an issue that block could be produced before completion of chunk in some rare cases. Because we process pending endorsements only on completion, we were never taking them into account then.

Processing pending endorsements before making chunk ready for inclusion allows to verify that cluster of 4 nodes can generate 12 blocks full of chunks.

https://nayduck.nearone.org/#/run/120, look at `chunks_management` tests. Moving them to `expensive` because much more network is involved.

### CODE REVIEW

**Comment by Longarithm** [comment]
Actually https://nayduck.nearone.org/#/run/120 shows that tests are flaky, and I could reproduce some failure locally after 3 tries.

**Comment by Longarithm** [comment]
Let's just merge the fix for now?

File: integration-tests/src/tests/client/chunks_management.rs [hunk 1]
```diff
@@ -44,15 +42,23 @@ struct Test {
 impl Test {
     fn run(self) {
         // TODO(#10506): Fix test to handle stateless validation
-        if checked_feature!("stable", StatelessValidationV0, PROTOCOL_VERSION) {
+        if near_primitives_core::checked_feature!(
+            "stable",
+            StatelessValidationV0,
+            near_primitives_core::version::PROTOCOL_VERSION
+        ) {
```
[the remark this answers is not in the document]
Comment (tayfunelmas) [reply integration-tests/src/tests/client/chunks_management.rs:49]: oh ok I see the comment about flakiness.

**Review by tayfunelmas — APPROVED** [review]

**Review by bowenwang1996 — COMMENTED** [review]
looks like `self.chunk_endorsement_tracker.process_pending_endorsements(&chunk_header);` is always called before ` self.chunk_inclusion_tracker
            .mark_chunk_header_ready_for_inclusion(chunk_header, validator_id);` Should we consider refactoring those two lines into a separate function?

**Comment by tayfunelmas** [comment]
Based on [this message](https://near.zulipchat.com/#narrow/stream/407237-core.2Fstateless-validation/topic/.E2.9C.94.20chunk.20completion.20not.20guaranteed.20on.20BPs/near/441469132) the issue comes from the fact that in some cases `on_chunk_completed` is not called, otherwise it would lead to the same function you are calling here (`process_pending_endorsements`). Is it the case that in those scenarios `on_chunk_completed` not called at all, or called later? Is there a way to make sure `ShardsManagerResponse::ChunkCompleted` is generated and `on_chunk_completed` is called instead of calling the same function for a different event (`ShardsManagerResponse::ChunkHeaderReadyForInclusion`)?

**Review by shreyan-gupta — APPROVED** [review]
I think it makes sense to remove the call to process_pending_endorsements in `on_chunk_completed` now? That's only a subset of cases and can make the code overall more confusing.

Also, might make sense to have a tiny comment explaining why we have the call to process_pending_endorsements here?

**Comment by Longarithm** [comment]
> Is it the case that in those scenarios on_chunk_completed not called at all, or called later?

It is called later, and it can be too late for block production.
Based on the code, unfortunately for different nodes `ChunkCompleted` means different things. In particular, if node doesn't track shard, it's enough to get some subset of parts and receipts; if it track shard, it means that chunk can be reconstructed. So it is not safe to use that message. We would have needed to split it into two, e.g. `ChunkCompleted` and `ChunkConstructed`.

> I think it makes sense to remove the call to process_pending_endorsements in on_chunk_completed now?

Could be and tests seem to pass, but it looks that in `try_process_chunk_parts_and_receipts`, `can_reconstruct` can be true while `have_all_parts` is false - like, when node received big enough subset of parts to reconstruct chunk, but not all of parts it owns. But if node constructed the chunk, it can compute any part, so why not to process pending endorsements... So this is a bit subtle and I would consider it separately.

**Pull request merged**